ambry-store/src/test/java/com.github.ambry.store/IndexTest.java

```diff
@@ -697,21 +700,79 @@ public void findEntriesSinceOnRestartTest() throws IOException, StoreException {
+    startToken = new StoreFindToken(lastRecordOffset, oldSessionId, incarnationId, false);
+    expectedEndToken = new StoreFindToken(firstRecordFileSpan.getStartOffset(), sessionId, incarnationId, false);
     expectedEndToken.setBytesRead(bytesRead);
     doFindEntriesSinceTest(startToken, Long.MAX_VALUE, Collections.singleton(newId), expectedEndToken);
   }
 
   /**
+   * Tests behaviour of {@link PersistentIndex#findEntriesSince(FindToken, long)} on restart of index with a corner case
+   * scenario
+   * @throws IOException
+   * @throws StoreException
+   */
+  @Test
+  public void findEntriesSinceOnRestartCornerCaseTest() throws IOException, StoreException {
+    UUID oldSessionId = sessionId;
+    // ensure that index and log end offsets are in sync
+    addPutEntries(3, PUT_RECORD_SIZE, Utils.Infinite_Time);
+    // test case description:
+    // after restart, the store(no new writes have gone into the store) will return a journal based token referring to
+    // logEndOffsetOnStartUp for a findEntriesSince(offset beyond logEndOffsetOnStartup with different session id) call
+    // On the subsequent findEntriesSince() call, the index should start returning entries
+    // starting from that offset and should not consider it as non-inclusive offset
+    recovery = new MessageStoreRecovery() {
+      @Override
+      public List<MessageInfo> recover(Read read, long startOffset, long endOffset, StoreKeyFactory factory)
+          throws IOException {
+        return Collections.emptyList();
+      }
+    };
+    reloadIndex(true);
+    Offset lastRecordOffset = index.journal.getLastOffset();
+    long bytesRead =
+        log.getDifference(new Offset(log.getEndOffset().getName(), lastRecordOffset.getOffset() + PUT_RECORD_SIZE),
+            new Offset(log.getFirstSegment().getName(), 0));
+    // create a token that will be past the index end offset on start up after restart.
+    StoreFindToken startToken = new StoreFindToken(
+        new Offset(log.getEndOffset().getName(), (lastRecordOffset.getOffset() + (2 * PUT_RECORD_SIZE))), oldSessionId,
+        incarnationId, false);
+    // end token should point to end offset of last entry i.e. (lastRecordOffset + PUT_RECORD_SIZE)
+    StoreFindToken expectedEndToken =
+        new StoreFindToken(new Offset(log.getEndOffset().getName(), lastRecordOffset.getOffset() + PUT_RECORD_SIZE),
+            sessionId, incarnationId, true);
+    expectedEndToken.setBytesRead(bytesRead);
+    // Fetch the FindToken returned from findEntriesSince
+    FindInfo findInfo = index.findEntriesSince(startToken, Long.MAX_VALUE);
+    assertEquals("EndToken mismatch ", expectedEndToken, findInfo.getFindToken());
+    assertEquals("No entries should have been returned ", 0, findInfo.getMessageEntries().size());
+
+    // add 2 entries to index and log
+    addPutEntries(1, PUT_RECORD_SIZE, Utils.Infinite_Time);
+    Pair<MockId, IndexValue> indexEntry1 = logOrder.lastEntry().getValue();
```

Comment:
minor: these variables don't look very useful. Why not just create `expectedEntries` before and add here instead of using temp variables that serve no other purpose?